core/storage/trie/polkadot_trie/polkadot_trie_impl.cpp

```diff
@@ -276,176 +284,181 @@ namespace kagome::storage::trie {
-    if (not prefix_equal) {
-      return {parent, count};
+    if (not std::equal(parent->key_nibbles.begin(),
+                       parent->key_nibbles.end(),
+                       prefix.begin())) {
+      return outcome::success();
     }
-    using T = PolkadotNode::Type;
-    if (parent->getTrieType() == T::BranchWithValue
-        or parent->getTrieType() == T::BranchEmptyValue) {
-      auto branch = std::dynamic_pointer_cast<BranchNode>(parent);
 
+    if (parent->isBranch()) {
       const auto length = parent->key_nibbles.size();
-      BOOST_ASSERT(
-          length == getCommonPrefixLength(parent->key_nibbles, prefix_nibbles));
-
-      OUTCOME_TRY(child, retrieveChild(branch, prefix_nibbles[length]));
-      if (child == nullptr) {
-        return {parent, count};
-      }
-      OUTCOME_TRY(
-          tup, detachNode(child, prefix_nibbles.subspan(length + 1), callback));
-      auto to_detach = branch->children.at(prefix_nibbles[length]);
-      branch->children.at(prefix_nibbles[length]) = std::get<0>(tup);
-
-      OUTCOME_TRY(size, notifyIsDetached(to_detach, callback));
-      return {branch, count + size};
+      auto &child =
+          dynamic_cast<BranchNode &>(*parent.get()).children.at(prefix[length]);
+
+      OUTCOME_TRY(detachNode(child,
+                             prefix.subspan(length + 1),
+                             limit,
+                             finished,
+                             count,
+                             callback,
+                             retrieveChild));
+      OUTCOME_TRY(handleDeletion(parent, retrieveChild));
     }
-    return {parent, count};
+    return outcome::success();
   }
 
-  outcome::result<size_t> PolkadotTrieImpl::notifyIsDetached(
-      const PolkadotTrie::NodePtr &node, const OnDetachCallback &callback) {
-    size_t count = 0;
-    if (node) {
-      if (node->isBranch()) {
-        auto branch = std::dynamic_pointer_cast<BranchNode>(node);
-        for (auto &child : branch->children) {
-          OUTCOME_TRY(size, notifyIsDetached(child, callback));
-          count += size;
-        }
-      }
-
-      auto key = PolkadotCodec::nibblesToKey(node->key_nibbles);
-      OUTCOME_TRY(callback(key, std::move(node->value)));
-      ++count;
-    }
-    return outcome::success(count);
+  outcome::result<void> notifyOnDetached(
+      PolkadotTrie::NodePtr &node,
+      const PolkadotTrie::OnDetachCallback &callback) {
+    auto key = PolkadotCodec::nibblesToKey(node->key_nibbles);
+    OUTCOME_TRY(callback(key, std::move(node->value)));
+    return outcome::success();
   }
 
   outcome::result<PolkadotTrie::NodePtr> PolkadotTrieImpl::retrieveChild(
       BranchPtr parent, uint8_t idx) const {
     return retrieve_child_(std::move(parent), idx);
   }
 
-  uint32_t PolkadotTrieImpl::getCommonPrefixLength(
-      const KeyNibbles &first, const KeyNibbles &second) const {
-    auto &&[it1, it2] =
+  uint32_t getCommonPrefixLength(const KeyNibbles &first,
```

Comment:
Please provide the docstring.

Reply:
Why not did you define free functions at the top of the file?
When they are not supposed to be exposed, just define somewhere at the beginning inside an anonymous namespace.